x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/RBACEngine.java

```diff
@@ -0,0 +1,349 @@
+                    listener::onFailure));
+            } else {
+                listener.onResponse(new RBACAuthorizationInfo(role, role));
+            }
+        }, listener::onFailure));
+    }
+
+    private void getRoles(User user, ActionListener<Role> roleActionListener) {
+        // we need to special case the internal users in this method, if we apply the anonymous roles to every user including these system
+        // user accounts then we run into the chance of a deadlock because then we need to get a role that we may be trying to get as the
+        // internal user. The SystemUser is special cased as it has special privileges to execute internal actions and should never be
+        // passed into this method. The XPackUser has the Superuser role and we can simply return that
+        if (SystemUser.is(user)) {
+            throw new IllegalArgumentException("the user [" + user.principal() + "] is the system user and we should never try to get its" +
+                " roles");
+        }
+        if (XPackUser.is(user)) {
+            assert XPackUser.INSTANCE.roles().length == 1;
+            roleActionListener.onResponse(XPackUser.ROLE);
+            return;
+        }
+        if (XPackSecurityUser.is(user)) {
+            roleActionListener.onResponse(ReservedRolesStore.SUPERUSER_ROLE);
+            return;
+        }
+
+        Set<String> roleNames = new HashSet<>(Arrays.asList(user.roles()));
+        if (isAnonymousEnabled && anonymousUser.equals(user) == false) {
+            if (anonymousUser.roles().length == 0) {
+                throw new IllegalStateException("anonymous is only enabled when the anonymous user has roles");
+            }
+            Collections.addAll(roleNames, anonymousUser.roles());
+        }
+
+        if (roleNames.isEmpty()) {
+            roleActionListener.onResponse(Role.EMPTY);
+        } else if (roleNames.contains(ReservedRolesStore.SUPERUSER_ROLE_DESCRIPTOR.getName())) {
+            roleActionListener.onResponse(ReservedRolesStore.SUPERUSER_ROLE);
+        } else {
+            rolesStore.roles(roleNames, fieldPermissionsCache, roleActionListener);
+        }
+    }
+
+    @Override
+    public void authorizeRunAs(Authentication authentication, TransportRequest request, String action, AuthorizationInfo authorizationInfo,
+                               ActionListener<AuthorizationResult> listener) {
+        if (authorizationInfo instanceof RBACAuthorizationInfo) {
+            final Role role = ((RBACAuthorizationInfo) authorizationInfo).getAuthenticatedUserAuthorizationInfo().getRole();
+            listener.onResponse(new AuthorizationResult(role.runAs().check(authentication.getUser().principal())));
+        } else {
+            listener.onFailure(new IllegalArgumentException("unsupported authorization info:" +
+                authorizationInfo.getClass().getSimpleName()));
+        }
```

Comment:
How about having generic `AuthorizationEngine<AuthorizationInfo>` to avoid these pesky conditions ?

Reply:
I spent time trying to use generics earlier and ran into issues in the authorization service IIRC; but I can give it another shot. However, I would like to keep that change as a followup to avoid the churn that might cause in reviewing this.

Reply:
> However, I would like to keep that change as a followup to avoid the churn that might cause in reviewing this.

Okay, that's best.